### ISSUE TITLE
Methodology for finding Egg-Room

### ISSUE THREAD

**Issue by hiburn8** [JayFoxRox/futurama-tools#1]
I'm really amazed you and the community were able to figure out how to open the egg room... but i'm just wondering exactly what your process was when looking through the scripts? What data were you looking at to make the decision that you needed to do A -> B -> C -> D? 
I'd love to try and replicate this for my own learning.

**Comment by JayFoxRox** [comment]
I don't fully remember my initial thought process, but most of it is probably covered in the README.
Take this with a grain of salt. If there's conflicting stories by me elsewhere, trust the older story - this is written from memory.

1. Saw the video and thought "this is interesting; this could be fun".
2. I believe I did a `grep egg -i` on all files to find possible mentions of it (I used the Xbox version, which luckily is grouping data - by type - into smaller IMG files than PS2).
3. Probably extracted the IMG files now (might have done for the previous step already?), using [Game Extractor](http://www.watto.org/game_extractor.html).
4. There were mention of "scripts" in the IMG file, and the .ful files looked the most likely probably (debug messages you'd find in code).
5. Started trying to find patterns in the ful files for a disassembler - this took a lot of time. I had some *rough* parser working, but I didn't know what most of the data meant (I probably still wasn't sure about this being a script).
6. Went back to look for other files, finding the .dbg files (or I had seen them before?), and decided to write a parser for those; what made those interesting is that they also had filepath to "/scripts/". At the time I was not sure if they had any relation to the .ful files (other than being in the same directory).
7. Once I had a parser, I recognized some connections between the file formats, and I merged my ful and dbg code to get a much more readable output. This was mostly function names (internal / external) from the dbg, because I still didn't know how variables worked (and to this day, I'm still not confident).
8. My initial assembly output still had various candidates for variable names / types on each script operand. I tried to find patterns and changed the code to pick the most likely candidate.
9. I mostly focused on understanding the `CirclePoint(FuPoint3 center, float radius, float angle)` function (and improved my made-up assembly-language), because I had a rough idea of what it would do. I have rewritten this function from assembly to high-level language to gain confidence and to verify my assembly-language design.
10. Eventually I reached a point where you could follow along the disassembly, this is where early https://gist.github.com/JayFoxRox/bc4f2e781428d69f591565e1dc545fa3/revisions comes from (the function name from the dbg made somewhat obvious which function to investigate).
11. With a lot of the rough ingredients, a friend of mine tried various ways to pick up the egg already - and failed. I also misunderstood a lot of things and thought that certain conditions were "shortcuts" to get the egg - I did not realize they were "do-not-do-this"-checks (the incorrect "short-cut" comments remain in my GitHub Gist).
12. Because it was obvious that there's a complicated set of steps (or bugs in my variable names; seeing how stuff like the fire extinguisher were mentiond), I rewrote the logic in a higher-level pseudo-language (block-by-block, not thinking too much about the bigger-picture).
13. Only once the entire high-level code was rewritten like that, I turned on my Xbox and went ingame.
14. Most steps were very clear, but there were 3 steps in particular that required a lot of trial-and-error:
    - `StageA.m_LabDrawer.m_CurState == 1`: Is this opened or closed? I basically just attempted everything I did twice: once with lab-drawer opened, once closed.
    - `StageA.m_Tv.m_State != 2`: What state is this? I believe I read the script disassembly for the TV code to figure this out.
    - `:FsGetCollPickedUp() != 7`: I believed the `Coll` stuff would be keycards and mission objects (fire-extinguisher, hammer, ..); I ran through the first level countless of times, counting *any* kind of object I had collected (including nibbler, cash, and other objects) - I also wasn't sure if I had to return the items back to the Professor to count, or if used-keycards still count to this value. It took a *long* time until I realized that *only* the money counted (which has an on-screen counter).
15. Once I confirmed my method to reach the egg, I also reviewed the code to make sure I didn't miss anything. I had written rough instructions along the way, and kept refining them (staying as close to the script as possible).

---

As for replicating:

If you want to start at step 10/12, read the source code of my tools ([like this](https://github.com/JayFoxRox/futurama-tools/blob/master/disassemble_ful.py#L302-L303)) to figure out how to use them.

If you want to start at step 1/2: Good luck! Feel free to diverge from my path - I think starting with dbg files would have been smarter; but I couldn't have known at that time. Decision for ful / dbg was purely based on a greedy search I think - and ful looked more interesting; dbg could have been outdated debug information - which wouldn't have helped at all, but wasted a lot of time.

---

As I was trying to reach oddheader, I was also told about other Futurama easter-eggs. So I wanted to confirm them independently by looking at the scripts.
I also started looking for differences in the PS2 and Xbox versions of the game-scripts; so I cleaned up a lot of my workflow / tooling, to reduce errors.

Also, as the scripts referenced trigger-zones, I also wanted a map viewer.
The nifskope based map viewer didn't take that long to develop actually. But it still isn't quite-stable yet and it also doesn't support animation playback / texture-viewing.
The script is also rather tightly integrated with the maps (scripts seem to connect animations to objects etc.), so the map viewer itself can't do certain things I'd like it to do.

Ideally, we'd also have a script-assembler, so we could try many of those unused script functions (there's *a lot*) and reach the unreachable code.
A macro-disassembler would also be helpful as there's a couple of repeating patterns.
However, writing these tools is not a priority for me.

---

During my Futurama work, I have independently confirmed all known Futurama easter-eggs; I also have discovered at least 1 egg, that isn't publicly documented (yet).
I also found (presumably) incomplete easter-egg code, unreachable code, and cut content (2 unused levels in the game; which don't work, but do load in the map viewer).
Most of this was done using `grep` to find interesting stuff; some interesting content was also discovered by other people (LarryLemmy was noticing a lot of interesting things lately).
I then made / refined my tools for a full analysis.

---

My plan is to accumulate all of the discovered information for a technical post [on my Patreon: https://www.patreon.com/JayFoxRox](https://www.patreon.com/jayfoxrox).

However, lately a lot of the "leaked" information has been independently found (although with errors) / documented [on TCRF](https://tcrf.net/Futurama#Easter_Eggs).
I hope once I finish that [Patreon](https://www.patreon.com/jayfoxrox) post, that a lot of the errors will go away, with my research becoming the primary source (over trial-and-error).

---

*(Feel free to close this issue once your question has been answered; if you have more questions, you can add questions in this topic)*

**Comment by hiburn8** [comment]
This is amazing, thank you so much Jay! Very impressive stuff... i'll definitely spend some time trying to replicate your efforts. And I'm so happy that you've been able to shine light on this game again. It was actually a surprise favourite of mine as a kid :)
